spec.html

```diff
@@ -23028,8 +23116,129 @@ <h1>InnerModuleEvaluation ( _module_, _stack_, _index_ )</h1>
+            <h1>GatherAsyncParentCompletions ( _module_, _execList_ )</h1>
+            <p>The abstract operation GatherAsyncParentCompletions takes arguments _module_ (a Cyclic Module Record) and _execList_ (a List of Cyclic Module Records). It performs the following steps when called:</p>
+            <emu-alg>
+              1. For each Cyclic Module Record _m_ of _module_.[[AsyncParentModules]], do
+                1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
+                  1. Assert: _m_.[[Status]] is ~evaluating-async~.
+                  1. Assert: _m_.[[EvaluationError]] is ~empty~.
+                  1. Assert: _m_.[[AsyncEvaluation]] is *true*.
+                  1. Assert: _m_.[[PendingAsyncDependencies]] &gt; 0.
+                  1. Set _m_.[[PendingAsyncDependencies]] to _m_.[[PendingAsyncDependencies]] - 1.
+                  1. If _m_.[[PendingAsyncDependencies]] = 0, then
+                    1. Append _m_ to _execList_.
+                    1. If _m_.[[Async]] is *false*, perform ! GatherAsyncParentCompletions(_m_, _execList_).
+              1. Return *undefined*.
+            </emu-alg>
+            <emu-note>
+              <p>When an async execution for a root _module_ is fulfilled, this function determines the list of modules which are able to synchronously execute together on this completion, populating them in _execList_.</p>
+            </emu-note>
+          </emu-clause>
+
+          <emu-clause id="sec-async-module-execution-fulfilled" aoid="AsyncModuleExecutionFulfilled">
+            <h1>AsyncModuleExecutionFulfilled ( _module_ )</h1>
+            <p>The abstract operation AsyncModuleExecutionFulfilled takes argument _module_ (a Cyclic Module Record). It performs the following steps when called:</p>
+            <emu-alg>
+              1. If _module_.[[Status]] is ~evaluated~, then
+                1. Assert: _module_.[[EvaluationError]] is not ~empty~.
+                1. Return *undefined*.
+              1. Assert: _module_.[[Status]] is ~evaluating-async~.
+              1. Assert: _module_.[[AsyncEvaluation]] is *true*.
+              1. Assert: _module_.[[EvaluationError]] is ~empty~.
+              1. Set _module_.[[AsyncEvaluation]] to *false*.
+              1. Set _module_.[[Status]] to ~evaluated~.
+              1. If _module_.[[TopLevelCapability]] is not ~empty~, then
+                1. Assert: _module_.[[CycleRoot]] is _module_.
+                1. Perform ! Call(_module_.[[TopLevelCapability]].[[Resolve]], *undefined*, &laquo; *undefined* &raquo;).
+              1. Let _execList_ be a new empty List.
+              1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+              1. Let _sortedExecList_ be a List whose elements are the elements of _execList_, in the order in which they had their [[AsyncEvaluation]] fields set to *true* in InnerModuleEvaluation.
```

Comment:
I'm leery of this, as I'm pretty sure there's nowhere in the current spec that behavior depends on the order in which fields were set to some value.

Reply:
We discussed this briefly in the editor call today. I also don't love it, but haven't yet thought of an alternative I like more, and there's nothing technically wrong with this approach. (A global counter would work, for example, but isn't exactly cleaner.)

The closest thing we currently have is the step in [OrdinaryOwnPropertyKeys](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-ordinaryownpropertykeys) which reads

> For each own property key P of O such that Type(P) is String and P is not an array index, in ascending chronological order of property creation, do

Reply:
I prefer the current wording or something similar, as it gets the intention across more clearly. I agree -- it is the first time we do this and i was a bit worried about it when we first introduced it. I am open to ways we might do this while still maintaining a clear model for readers.

Reply:
Yes, the main reason I'm okay with this current wording is because as @bakkot says: an explicit "implementation" of the intent via e.g. a global counter just isn't any clearer.